### PR TITLE
Implementations of Copy file to container command

### PR DESCRIPTION
**Pull request by vuminhkh**
Hello,

I've implemented "Copy file to container command", but it seems that migrating to 1.20 API breaks other part (8 tests failure). The commands that I modified, the test works well.

As I don't have enough knowledge of the project to fix them all, I still perform the pull request, hoping that it will be useful.

Thanks,

Minh Khang VU

### CODE REVIEW

**Comment by marcuslinke** [comment]
@vuminhkh Thanks for your contribution. I will merge ASAP.

File: src/main/java/com/github/dockerjava/core/CompressArchiveUtil.java [hunk 1]
```diff
@@ -2,18 +2,77 @@
...
 
+import com.google.common.io.ByteStreams;
+import com.google.common.io.Closeables;
+
 public class CompressArchiveUtil {
 
+    static void putTarEntry(TarArchiveOutputStream tarOutputStream, TarArchiveEntry tarEntry, Path file) throws IOException {
+        tarEntry.setSize(Files.size(file));
+        tarOutputStream.putArchiveEntry(tarEntry);
+        InputStream input = new BufferedInputStream(Files.newInputStream(file));
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/CompressArchiveUtil.java]: Place in try-with-resources?

File: src/main/java/com/github/dockerjava/core/DockerClientImpl.java [hunk 2]
```diff
@@ -1,5 +1,12 @@
 package com.github.dockerjava.core;
 
+import static com.google.common.base.Preconditions.checkNotNull;
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/DockerClientImpl.java]: static usually imported after all imports

File: src/main/java/com/github/dockerjava/core/DockerClientImpl.java [hunk 3]
```diff
@@ -77,16 +86,8 @@
 import com.github.dockerjava.core.command.VersionCmdImpl;
 import com.github.dockerjava.core.command.WaitContainerCmdImpl;
 
-import java.io.Closeable;
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/DockerClientImpl.java]: why do you reorder imports, what IDE do you use?

**Comment by KostyaSha** [comment]
I see many unclosed streams, @vuminhkh could you check that you have no resource leaks https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html ?

File: src/main/java/com/github/dockerjava/core/CompressArchiveUtil.java (hunk 1 above)
Comment (vuminhkh) [inline src/main/java/com/github/dockerjava/core/CompressArchiveUtil.java]: Thank you. It's better to go with new Java 7 functionality. It's cleaner.

File: src/main/java/com/github/dockerjava/core/DockerClientImpl.java (hunk 2 above)
Comment (vuminhkh) [inline src/main/java/com/github/dockerjava/core/DockerClientImpl.java]: I'm using Intellij, it's re-arranged automatically like that by default. Please feel free to re-organize.

File: src/main/java/com/github/dockerjava/core/DockerClientImpl.java (hunk 3 above)
Comment (vuminhkh) [inline src/main/java/com/github/dockerjava/core/DockerClientImpl.java]: I'm using Intellij, it's re-arranged automatically like that by default. Please feel free to re-organize.

**Comment by vuminhkh** [comment]
Hello,

For streams unclosed, if i'm not mistaken it's about the tar method in CompressArchiveUtil, as it's stream which wrap stream, closing the most outer will delegate to inner and the whole stream wrapped chain will be closed.

For formatting rules, i'm using Intellij, please feel free to re-organize based on your IDE settings, or give me some hints to have the same format rules as yours.

Thanks,

Minh Khang VU

**Comment by marcuslinke** [comment]
@vuminhkh After thinking about again we should stay backward compatible. In fact the `CopyFileFromContainerCmd` should target to `/containers/{id}/copy` and then
a new `CopyArchiveFromContainerCmd` as well as a `CopyArchiveToContainerCmd` should target to `
/containers/{id}/archive`. Would it be possible for you to change your PR accordingly please?

**Comment by vuminhkh** [comment]
Hi Marcus,

I'll find my-self some times to do this asap.

Minh Khang VU

**Comment by vuminhkh** [comment]
Hi Marcus,

Please review to see if it's ok. There are always 8 test failures (the same as before) but they don't concern the commands that I added/modified.

Regards,

Minh Khang VU
[TestSuite.txt](https://github.com/docker-java/docker-java/files/42201/TestSuite.txt)

File: src/main/java/com/github/dockerjava/api/command/CopyArchiveFromContainerCmd.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     *
+     * @throws NotFoundException
+     *             No such container
+     */
+    @Override
+    public InputStream exec() throws NotFoundException;
+
+    public static interface Exec extends DockerCmdSyncExec<CopyArchiveFromContainerCmd, InputStream> {
+    }
+}
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/command/CopyArchiveFromContainerCmd.java]: new line in the end of file

File: src/main/java/com/github/dockerjava/api/DockerClient.java [hunk 5]
```diff
@@ -119,8 +121,12 @@
 
     public LogContainerCmd logContainerCmd(String containerId);
 
+    public CopyArchiveFromContainerCmd copyArchiveFromContainerCmd(String containerId, String resource);
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/DockerClient.java]: please add javadoc and mention from what docker-api version it works.

File: src/main/java/com/github/dockerjava/api/command/CopyArchiveToContainerCmd.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    public boolean isNoOverwriteDirNonDir();
+
+    public boolean isDirChildrenOnly();
+
+    /**
+     * Set container's id
+     *
+     * @param containerId id of the container to copy file to
+     * @return this
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/command/CopyArchiveToContainerCmd.java]: remove return description, it makes no sense

File: src/main/java/com/github/dockerjava/core/command/CopyFileFromContainerCmdImpl.java [hunk 7]
```diff
@@ -74,4 +74,4 @@ public CopyFileFromContainerCmdImpl withHostPath(String hostPath) {
     public InputStream exec() throws NotFoundException {
         return super.exec();
     }
-}
+}
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/command/CopyFileFromContainerCmdImpl.java]: unrelated to PR change

File: src/main/java/com/github/dockerjava/core/util/FilePathUtil.java [hunk 8]
```diff
@@ -12,11 +13,9 @@ private FilePathUtil() {
 
     /**
      * Return the relative path. Path elements are separated with / char.
-     * 
-     * @param baseDir
-     *            a parent directory of {@code file}
-     * @param file
-     *            the file to get the relative path
+     *
+     * @param baseDir a parent directory of {@code file}
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/util/FilePathUtil.java]: unrelated change

**Comment by KostyaSha** [comment]
@vuminhkh please exclude all unrelated to your change formatting and import changes

**Comment by vuminhkh** [comment]
@KostyaSha I tried to use your formatting rules at docker-java/etc/code-style.epf and eclipse plugin for intellij, it seems that i's not your current formatting rules. So I made all the changes "manually", what do you advice about formatting rules in order to contribute to your project ?

**Comment by marcuslinke** [comment]
@vuminhkh Thanks! Hopefully I'm able to merge this evening...

**Comment by marcuslinke** [comment]
@vuminhkh After looking deeper into this I've just found that the API of `CopyArchiveToContainerCmd` prevents from copying an already compressed tar archive to the container. This is useful if someone wants to copy an archive from one container to another one.

The command should be able to transport a `InputStream` to the `CopyArchiveToContainerCmdExec` just like the `BuildImageCmd` does. The tar compression must be done in the command and not in the exec class then. Would it be possible for you to adopt your PR accordingly please?

Beside this, the NullpointerExceptions that you got are caused by https://github.com/docker-java/docker-java/issues/388

File: src/main/java/com/github/dockerjava/api/DockerClient.java [hunk 9]
```diff
@@ -119,8 +121,37 @@
...
+     *
+     * @param containerId id of the container
+     * @param resource    path to container's resource
+     * @return created command
+     */
+    public CopyArchiveFromContainerCmd copyArchiveFromContainerCmd(String containerId, String resource);
+
+    /**
+     * Copy resource from container to local machine.
+     * Deprecated, since docker API version 1.20 and above use copyArchiveFromContainerCmd instead.
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/DockerClient.java]: `@deprecated` javadoc

File: src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecFactoryImpl.java [hunk 10]
```diff
@@ -173,7 +177,7 @@ public void init(DockerClientConfig dockerClientConfig) {
     }
 
     private org.apache.http.config.Registry<ConnectionSocketFactory> getSchemeRegistry(final URI originalUri,
-            SSLContext sslContext) {
+                                                                                       SSLContext sslContext) {
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecFactoryImpl.java]: unrelated change

File: src/test/java/com/github/dockerjava/core/command/CopyFileFromContainerCmdImplTest.java [hunk 11]
```diff
@@ -70,4 +70,4 @@ public void copyFromNonExistingContainer() throws Exception {
         } catch (NotFoundException ignored) {
         }
     }
-}
+}
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/command/CopyFileFromContainerCmdImplTest.java]: unrelated change

**Comment by vuminhkh** [comment]
@marcuslinke I agree with you, It's not much modification but the API is more generic. I updated the API according to your suggestion.
@KostyaSha API doc and deprecated annotation updated to Java standard. For indentation issues, I insist that you commit your formatting rules and I'll use it to reformat files that I modified.

Thanks,

Minh Khang VU

**Comment by KostyaSha** [comment]
@vuminhkh i slowly adding checkstyle that matches to what marcus did. The standard rule for projects that has not yet defined styles is not mix styles and follow existed.

File: src/main/java/com/github/dockerjava/api/DockerClient.java [hunk 12]
```diff
@@ -121,8 +123,37 @@
...
+    public CopyArchiveFromContainerCmd copyArchiveFromContainerCmd(@Nonnull String containerId, @Nonnull String resource);
+
+    /**
+     * Copy resource from container to local machine.
+     *
+     * @param containerId id of the container
+     * @param resource    path to container's resource
+     * @return created command
+     * @see #copyArchiveFromContainerCmd(String, String)
+     * @deprecated since docker API version 1.20, replaced by {@link #copyArchiveFromContainerCmd(String, String)}
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/DockerClient.java]: (Not critical, but for simplification there are API versions for linking javadocs. Example https://github.com/docker-java/docker-java/blob/58d94079d2cd7ae7d072e5aad599777707e5b10a/src/main/java/com/github/dockerjava/api/command/InspectContainerResponse.java#L234 )

**Comment by vuminhkh** [comment]
@KostyaSha updated javadoc following your hint.
@marcuslinke and @KostyaSha I began to implement new network API, you can preview at https://github.com/vuminhkh/docker-java/tree/network . I'll perform another pull request once this one merged.

Thanks,

Minh Khang VU

**Comment by KostyaSha** [comment]
@vuminhkh implement classes is pretty simple, problem is wrote tests and verify this changes.

File: src/main/java/com/github/dockerjava/api/DockerClient.java (hunk 12 above)
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/DockerClient.java]: Try change link to version here also

**Comment by vuminhkh** [comment]
@KostyaSha For commands that I implemented, if i'm not mistaken, I've written tests for them all. Normally what I wrote did not break any existing feature, tests that are broken were APIs which are not compatible anymore with newer Docker API. I have not fixed those tests yet, I'll do it once I have more time. If you guy could help me also on this it will be great as those I don't have knowledge about those existing command.

**Comment by vuminhkh** [comment]
@marcuslinke @KostyaSha For breaking changes in docker API, what will be your strategy to be compatible with multiple docker api version ?

**Comment by KostyaSha** [comment]
Depends on cases.

2015-12-05 21:45 GMT+03:00 vuminhkh <notifications@github.com>:

**Comment by vuminhkh** [comment]
@KostyaSha What is the configuration of your test ? Which version of docker do you target to assure that there's not regression ? If i'm not mistaken it's here https://github.com/docker-java/docker-java/blob/master/src/test/java/com/github/dockerjava/client/AbstractDockerClientTest.java#L41 that you define which version of API the test target. Is it ok if I change it to 1.21 ?

**Comment by KostyaSha** [comment]
I'm bit newbie with current tests design. @marcuslinke please suggest.

**Comment by marcuslinke** [comment]
@vuminhkh, @KostyaSha  Currently we can only test against one specific version of the remote API. This is also the reason for the compatibility statement here: https://github.com/docker-java/docker-java#docker-java

As you already found this is hard-coded in the `AbstractDockerClientTest` class. I will push some changes related to issue https://github.com/docker-java/docker-java/issues/388 soon that will change the tested API version.

**Comment by marcuslinke** [comment]
After merging https://github.com/docker-java/docker-java/pull/392 I will take a look at your latest changes @vuminhkh.

**Comment by marcuslinke** [comment]
@vuminhkh Thanks for contributing!
